### PR TITLE
Clean up keyboard navigation for the inserter.

### PR DESCRIPTION
**Pull request by BE-Webdesign**
Cleaning up keyboard navigation for the inserter.  Some micro
optimizations and code pattern changes.  Overall cleanup.  Moving the
findNext() findPrevious() to use a common findByIncrement function.
Fixes broken sorting for blocks as well which caused odd behavior while
navigating via keyboard.

**Testing Instructions:**

Verify that the keyboard navigation for the inserter functions properly.  Currently it will skip pullquote until the end of the cycle.  Make sure this is not the case anymore.

### CODE REVIEW

**Comment by BE-Webdesign** [comment]
@aduth, I made all of the changes you requested minus one.  I still don't understand why we would want to call `getCategories()`, `getBlocks()` over and over rather than assigning them as a value to the constructor.  Are these values expected to change at some point while the editor is running? If so, what use case does that provide?

**Comment by aduth** [comment]
>I still don't understand why we would want to call `getCategories()`, `getBlocks()` over and over rather than assigning them as a value to the constructor.

The main concern stems from the fact that in doing so, you're creating a copy of the source of truth. You're correct to note that this is mainly an issue when the values are known to change, at which point keeping them in sync is cumbersome, error-prone, and verbose; and doing so incorrectly can lead to subtle bugs that are difficult to track down.

>Are these values expected to change at some point while the editor is running?

Right now, it wouldn't. Whether we expect this to always be the case, I can't say with certainty. Doesn't seem too unreasonable to want to allow blocks to be registered or unregistered on-demand, though.

Rather, I'd phrase this in terms of: What are you trying to achieve by creating a copy? Typically I see this as a cache of sorts, with the assumption that the underlying data is slow to generate. To this I'd say:

- Have we measured the impact? Adding complexity for a negligible difference is not worth the overhead cost
- The implementations of [`getCategories`](https://github.com/WordPress/gutenberg/blob/90cc25b78ca4df50212b9e1b06a8e31dcfc8550d/blocks/api/categories.js#L15-L22) and [`getBlocks`](https://github.com/WordPress/gutenberg/blob/e05d44dd51e1511610fef11efca85905489714c9/blocks/api/registration.js#L99-L106) are in fact not heavy
- Even if they were, is caching at the component level the best solution? What if multiple components operate on this data? If needed, we could easily memoize the underlying functions themselves to guarantee caching across the board (see [Lodash's `_.memoize`](https://lodash.com/docs/4.17.4#memoize))

**Comment by BE-Webdesign** [comment]
@aduth 

Changed it!  Let's merge?  This fixes the bug on sorting which makes navigating to the pullquote work properly.

File: editor/components/inserter/menu.js [hunk 1]
```diff
@@ -119,140 +84,127 @@ class InserterMenu extends wp.element.Component {
...
-			return blockTypes[ 0 ].slug;
-		}
+	findByIncrement( blockTypes, increment = 1 ) {
+		// Add on a fake search block to the list to cycle through.
+		const list = blockTypes.concat( { slug: 'search' } );
 
-		const currentIndex = blockTypes.findIndex( ( blockType ) => currentBlock === blockType.slug );
-		const nextIndex = currentIndex + 1;
-		const highestIndex = blockTypes.length - 1;
+		const currentIndex = findIndex( list, ( blockType ) => this.state.currentFocus === blockType.slug );
```
Comment (aduth) [inline editor/components/inserter/menu.js]: No need to change, but Lodash's shorthands (which are unfortunately slated to be disabled-by-default in the [upcoming major version](https://github.com/lodash/lodash/wiki/Roadmap)) are a nice simplification.

Before:

```js
findIndex( list, ( blockType ) => this.state.currentFocus === blockType.slug )
```

After:

```js
findIndex( list, { slug: this.state.currentFocus } )
```

**Review by aduth — APPROVED** [review]
Nice to see the sea of red; twice as many lines removed as there are modified! Good improvement 👍

File: editor/components/inserter/menu.js (hunk 1 above)
Comment (BE-Webdesign) [inline editor/components/inserter/menu.js]: Good to know.  That is somewhat hard to read, especially if you didn't know lodash could do that.  Did not know about lodash.memoize either.  Thank you for the knowledge dropping.

**Comment by aduth** [comment]
Related: #578

**Pull request merged**